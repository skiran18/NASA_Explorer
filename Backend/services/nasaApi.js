const axios = require('axios');
const config = require('../config');

const nasaClient = axios.create({
  baseURL: config.nasa.baseUrl,
  params: {
    api_key: config.nasa.apiKey
  }
});

class NasaApiService {
  async getAPOD() {
    const response = await nasaClient.get('/planetary/apod');
    return response.data;
  }

  async getMarsRoverPhotos(params) {
    console.log(params)
    const response = await nasaClient.get('/mars-photos/api/v1/rovers/curiosity/photos', {
      params
    });
    return response.data;
  }

  async getNearEarthObjects(params) {
    const response = await nasaClient.get('/neo/rest/v1/feed',{
      params
    });

    console.log(response)
    return response.data;
  }
}

module.exports = new NasaApiService();