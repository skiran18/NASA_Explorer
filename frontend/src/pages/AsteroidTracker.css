/* General Container */
.neo-container {
    max-width: 90%;
    margin: 50px auto;
    text-align: center;
    color: white;
  }
  
  /* Chart Section */
  .chart-section {
    background-color: #222;
    padding: 20px;
    border-radius: 8px;
    box-shadow: 0 4px 6px rgba(0, 0, 0, 0.4);
  }
  
  .chart-title {
    font-size: 24px;
    font-weight: bold;
    margin-bottom: 15px;
  }
  
  .chart-wrapper {
    height: 400px;
  }
  
  /* Asteroid Grid */
  .asteroid-grid {
    display: grid;
    grid-template-columns: repeat(auto-fit, minmax(280px, 1fr));
    gap: 20px;
    margin-top: 30px;
  }
  
  /* Asteroid Card */
  .asteroid-card {
    background-color: rgba(255, 255, 255, 0.1);
    padding: 15px;
    border-radius: 10px;
    transition: transform 0.3s ease-in-out;
    box-shadow: 0px 4px 6px rgba(0, 0, 0, 0.4);
  }
  
  .asteroid-card:hover {
    transform: scale(1.05);
  }
  
  /* Hazardous Asteroid */
  .hazardous {
    background-color: rgba(255, 0, 0, 0.3);
    border-left: 5px solid red;
  }
  
  /* Asteroid Header */
  .asteroid-header {
    display: flex;
    justify-content: space-between;
    align-items: center;
    margin-bottom: 10px;
  }
  
  .asteroid-name {
    font-size: 18px;
    font-weight: bold;
  }
  
  /* Hazard Icon */
  .hazard-icon {
    width: 24px;
    height: 24px;
    color: red;
  }
  
  /* Asteroid Info */
  .asteroid-info {
    text-align: left;
    font-size: 14px;
  }
  
  /* Loading Spinner */
  .loading-container {
    display: flex;
    justify-content: center;
    align-items: center;
    height: 400px;
  }
  
  .spinner {
    width: 50px;
    height: 50px;
    border: 5px solid rgba(255, 255, 255, 0.3);
    border-top: 5px solid #ffcc00;
    border-radius: 50%;
    animation: spin 1s linear infinite;
  }
  
  @keyframes spin {
    from {
      transform: rotate(0deg);
    }
    to {
      transform: rotate(360deg);
    }
  }
  
  /* Error Message */
  .error-message {
    font-size: 16px;
    color: red;
    margin-top: 20px;
  }
  