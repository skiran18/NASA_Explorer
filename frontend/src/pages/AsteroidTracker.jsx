import React from "react";

const AsteroidTracker = () => {
  return (
    <div className="asteroid-tracker">
      <h1>Near Earth Objects</h1>
      <p>Interactive 3D map and asteroid tracking data.</p>
    </div>
  );
};

export default AsteroidTracker;
