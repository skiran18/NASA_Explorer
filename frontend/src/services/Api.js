const API_BASE_URL = process.env.REACT_APP_API_BASE_URL;

export const fetchData = async (endpoint) => {
  try {
    const response = await fetch(`${API_BASE_URL}/${endpoint}`);
    console.log(response)
    if (!response.ok) {
      throw new Error(`Error: ${response.statusText}`);
    }
    return await response.json();
  } catch (error) {
    console.error("API Fetch Error:", error);
    throw error;
  }
};