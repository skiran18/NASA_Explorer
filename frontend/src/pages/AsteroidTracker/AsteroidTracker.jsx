import React, { useState, useEffect } from "react";
import {
  BarChart,
  Bar,
  XAxis,
  YAxis,
  Tooltip,
  ResponsiveContainer,
} from "recharts";
import { AlertTriangle } from "lucide-react";
import { fetchData } from "../../services/Api";
import "./AsteroidTracker.css";

const AsteroidTracker = () => {
  const [neoData, setNeoData] = useState();
  const [loading, setLoading] = useState(true);

  useEffect(() => {
    setLoading(true);
    fetchData("neo")
      .then((response) => {
        console.log(response);
        setNeoData(response);
        setLoading(false);
      })
      .catch((error) => {
        console.error("Error fetching NEO data:", error);
        setLoading(false);
      });
  },[]);

  if (loading) {
    return (
      <div className="loading-container">
        <div className="spinner"></div>
      </div>
    );
  }

  if (!neoData) {
    return <div className="error-message">Error loading NEO data</div>;
  }

  console.log(neoData)
   

  const now = new Date();
  const todayDate = now.toISOString().split("T")[0];

  const chartData = [
    {
      date: todayDate,
      count: neoData.near_earth_objects[todayDate].length,
      hazardous: neoData.near_earth_objects[todayDate].filter(
        (a) => a.is_potentially_hazardous_asteroid
      ).length,
    },
  ];

  return (
    <div className="neo-container">
      {/* Bar Chart Section */}
      <div className="chart-section">
        <h2 className="chart-title">Near Earth Objects Today</h2>
        <div className="chart-wrapper">
          <ResponsiveContainer width="100%" height="100%">
            <BarChart data={chartData}>
              <XAxis dataKey="date" />
              <YAxis />
              <Tooltip />
              <Bar dataKey="count" fill="#3b82f6" name="Total NEOs" />
              <Bar dataKey="hazardous" fill="#ef4444" name="Hazardous" />
            </BarChart>
          </ResponsiveContainer>
        </div>
      </div>

      {/* Asteroid Details Grid */}
      <div className="asteroid-grid">
        {Object.values(neoData.near_earth_objects)
          .flat()
          .map((asteroid) => (
            <div
              key={asteroid.id}
              className={`asteroid-card ${
                asteroid.is_potentially_hazardous_asteroid ? "hazardous" : ""
              }`}
            >
              <div className="asteroid-header">
                <h3 className="asteroid-name">{asteroid.name}</h3>
                {asteroid.is_potentially_hazardous_asteroid && (
                  <AlertTriangle className="hazard-icon" />
                )}
              </div>
              <div className="asteroid-info">
                <p>
                  <strong>Magnitude:</strong>{" "}
                  {asteroid.absolute_magnitude_h.toFixed(2)}
                </p>
                <p>
                  <strong>Miss Distance:</strong>{" "}
                  {parseInt(
                    asteroid.close_approach_data[0].miss_distance.kilometers
                  ).toLocaleString()}{" "}
                  km
                </p>
                <p>
                  <strong>Velocity:</strong>{" "}
                  {parseInt(
                    asteroid.close_approach_data[0].relative_velocity
                      .kilometers_per_hour
                  ).toLocaleString()}{" "}
                  km/h
                </p>
              </div>
            </div>
          ))}
      </div>
    </div>
  );
};

export default AsteroidTracker;
