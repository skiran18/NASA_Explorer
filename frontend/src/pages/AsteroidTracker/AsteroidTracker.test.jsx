import React from "react";
import { render, screen, waitFor } from "@testing-library/react";
import AsteroidTracker from "./AsteroidTracker";
import { fetchData } from "../../services/Api";
import "@testing-library/jest-dom";

// Mock fetchData to return test data
jest.mock("../../services/Api", () => ({
  fetchData: jest.fn(),
}));

beforeAll(() => {
    global.ResizeObserver = class {
      observe() {}
      unobserve() {}
      disconnect() {}
    };
  });

const mockData = {
  near_earth_objects: {
    "2025-02-15": [
      {
        id: "12345",
        name: "Asteroid XYZ",
        is_potentially_hazardous_asteroid: true,
        absolute_magnitude_h: 21.5,
        close_approach_data: [
          {
            miss_distance: { kilometers: "500000" },
            relative_velocity: { kilometers_per_hour: "25000" },
          },
        ],
      },
    ],
  },
};

describe("AsteroidTracker Component", () => {
  beforeEach(() => {
    fetchData.mockResolvedValue(mockData); // Mock API success response
  });

  jest.mock("../../services/Api", () => ({
    fetchData: jest.fn(() =>
      Promise.resolve({
        near_earth_objects: {
          "2025-02-15": [
            {
              id: "12345",
              name: "Asteroid XYZ",
              absolute_magnitude_h: 21.5,
              is_potentially_hazardous_asteroid: true,
              close_approach_data: [
                {
                  miss_distance: { kilometers: "500000" },
                  relative_velocity: { kilometers_per_hour: "25000" },
                },
              ],
            },
          ],
        },
      })
    ),
  }));
  
  test("AsteroidTracker Component matches snapshot", async () => {
    const { container } = render(<AsteroidTracker />);
    await waitFor(() => expect(screen.getByText("Near Earth Objects Today")).toBeInTheDocument());
    expect(container.firstChild).toMatchSnapshot();
  });

  test("renders the component with correct heading", async () => {
    render(<AsteroidTracker />);
    const headingElement = await screen.findByRole("heading", {
      name: /near earth objects/i,
    });

    expect(headingElement).toBeInTheDocument();
  });

test("AsteroidTracker Component matches snapshot", async () => {
    const { container } = render(<AsteroidTracker />);
  
    // ✅ Fix: Specifically wait for the <h2> heading
    await waitFor(() =>
      expect(screen.getByRole("heading", { level: 2 })).toBeInTheDocument()
    );
  
    // ✅ Fix: Handle multiple headings correctly
    const headings = screen.getAllByRole("heading");
    expect(headings.length).toBeGreaterThan(1); // Ensure multiple headings exist
  
    // Check for correct CSS class
    expect(container.firstChild).toHaveClass("neo-container");
  });

  test("matches snapshot", async () => {
    const { container } = render(<AsteroidTracker />);
  
    // Wait for the heading to appear (ensuring data has loaded)
    await waitFor(() =>
      expect(screen.getByRole("heading", { name: /near earth objects/i })).toBeInTheDocument()
    );
  
    // Now check the snapshot after the component has fully rendered
    expect(container.firstChild).toMatchSnapshot();
  });
});
