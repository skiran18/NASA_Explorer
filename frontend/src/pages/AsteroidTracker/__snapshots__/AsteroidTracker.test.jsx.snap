// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AsteroidTracker Component AsteroidTracker Component matches snapshot 1`] = `
<div
  class="neo-container"
>
  <div
    class="chart-section"
  >
    <h2
      class="chart-title"
    >
      Near Earth Objects Today
    </h2>
    <div
      class="chart-wrapper"
    >
      <div
        class="recharts-responsive-container"
        style="width: 100%; height: 100%; min-width: 0;"
      />
    </div>
  </div>
  <div
    class="asteroid-grid"
  >
    <div
      class="asteroid-card hazardous"
    >
      <div
        class="asteroid-header"
      >
        <h3
          class="asteroid-name"
        >
          Asteroid XYZ
        </h3>
        <svg
          class="hazard-icon"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m21.73 18-8-14a2 2 0 0 0-3.48 0l-8 14A2 2 0 0 0 4 21h16a2 2 0 0 0 1.73-3Z"
          />
          <path
            d="M12 9v4"
          />
          <path
            d="M12 17h.01"
          />
        </svg>
      </div>
      <div
        class="asteroid-info"
      >
        <p>
          <strong>
            Magnitude:
          </strong>
           
          21.50
        </p>
        <p>
          <strong>
            Miss Distance:
          </strong>
           
          5,00,000
           
          km
        </p>
        <p>
          <strong>
            Velocity:
          </strong>
           
          25,000
           
          km/h
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`AsteroidTracker Component matches snapshot 1`] = `
<div
  class="neo-container"
>
  <div
    class="chart-section"
  >
    <h2
      class="chart-title"
    >
      Near Earth Objects Today
    </h2>
    <div
      class="chart-wrapper"
    >
      <div
        class="recharts-responsive-container"
        style="width: 100%; height: 100%; min-width: 0;"
      />
    </div>
  </div>
  <div
    class="asteroid-grid"
  >
    <div
      class="asteroid-card hazardous"
    >
      <div
        class="asteroid-header"
      >
        <h3
          class="asteroid-name"
        >
          Asteroid XYZ
        </h3>
        <svg
          class="hazard-icon"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m21.73 18-8-14a2 2 0 0 0-3.48 0l-8 14A2 2 0 0 0 4 21h16a2 2 0 0 0 1.73-3Z"
          />
          <path
            d="M12 9v4"
          />
          <path
            d="M12 17h.01"
          />
        </svg>
      </div>
      <div
        class="asteroid-info"
      >
        <p>
          <strong>
            Magnitude:
          </strong>
           
          21.50
        </p>
        <p>
          <strong>
            Miss Distance:
          </strong>
           
          5,00,000
           
          km
        </p>
        <p>
          <strong>
            Velocity:
          </strong>
           
          25,000
           
          km/h
        </p>
      </div>
    </div>
  </div>
</div>
`;
